nanoFramework.Tools.DebugLibrary.Shared/PortSerial/PortSerial.cs

```diff
@@ -167,37 +147,66 @@ private void Device_ErrorReceived(object sender, SerialErrorReceivedEventArgs e)
         #endregion
 
         public ConnectPortResult ConnectDevice()
-        {
-            // try to determine if we already have this device opened.
+        {
+            ConnectPortResult openDeviceResult = ConnectPortResult.NotConnected;
+
+            // try to determine if we already have this device opened.
             if (Device != null &&
                 Device.IsOpen)
             {
                 // better make sure the RX FIFO it's cleared
                 Device.DiscardInBuffer();
 
                 return ConnectPortResult.Connected;
+            }
+
+            try
+            {
+                openDeviceResult = Policy.Handle<IOException>()
+                    .Or<UnauthorizedAccessException>()
+                    .Or<Exception>()
+                    .WaitAndRetry(10, retryCount => TimeSpan.FromMilliseconds((retryCount * retryCount) * 25),
+                        onRetry: (exception, delay, retryCount, context) => LogRetry(exception, delay, retryCount, context))
+                    .Execute(() => OpenDevice());
+
+                if (openDeviceResult == ConnectPortResult.Connected)
+                {
+                    OnLogMessageAvailable(NanoDevicesEventSource.Log.OpenDevice(InstanceId));
+                }
+                else 
+                {
+                    // Most likely the device is opened by another app, but cannot be sure
+                    OnLogMessageAvailable(NanoDevicesEventSource.Log.CriticalError($"Can't open Device: {InstanceId}"));
+                }
+            }
+            catch (UnauthorizedAccessException uaEx)
+            {
+                // Most likely the device is opened by another app, but cannot be sure
+                var logMsg = ($"Error in ConnectDevice: {uaEx.Message} Can't open {InstanceId}, possibly opened by another app");
+                Debug.WriteLine(logMsg);

```

Comment:
for the debug messages, either use the #if DEBUG or just remove them, it's a bit cleaner

Reply:
I believe Debug.WriteLine does the same as #if DEBUG.
But it's there because of the problem with OnLogMessageAvailable.  More specifically, the problem is the subsequent EventSource.WriteEvent. It stops writing when the thread is finished. And it comes to late. All is very complicated multithreading.

Reply:
I think @josesimoes  is an expert in this. He wrote an article about it on his blog:
https://jsimoesblog.wordpress.com/2020/05/12/debugger-output-in-c-sharp/